core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -116,14 +112,13 @@ public UfsStatus addStatus(AlluxioURI path, UfsStatus status) {
    * @return the previous set of children if the mapping existed, null otherwise
    */
   @Nullable
-  public Collection<UfsStatus> addChildren(AlluxioURI path, Collection<UfsStatus> children) {
-    Set<UfsStatus> set = new HashSet<>(children.size());
-    children.forEach(child -> {
+  public UfsStatus[] addChildren(AlluxioURI path, UfsStatus[] children) {
+    for (int i = 0; i < children.length; i++) {
+      UfsStatus child = children[i];
       AlluxioURI childPath = path.joinUnsafe(child.getName());
       addStatus(childPath, child);
-      set.add(child);
-    });
-    return mChildren.put(path, Collections.unmodifiableSet(set));
```

Comment:
This is the most significant change. I don't find a good way to make the array unmodifiable. The only good thing is, by currently using the unmodifiableSet, we guarantee existing callers don't modify the list, so changing to use array here doesn't break existing things. But if we want future guarantees, we might need to find another interface.

Reply:
Modifying array elements unfortunately can't be rejected like how unmodifiableSet does. But if we still want that attribute and don't want a copy by using existing data structures, we can easily write a wrapper like below to avoid copy in `new ArrayList(array)`:
```
class UnmodifiableList<T> implements List<T> {
  T[] elements;

  UnmodifiableList(T[] arr) {}
  @Override
  T get(int index) {}

  @Override
  void set(int index, T value) {throw new UnsupportedException();}
}
```

Reply:
On 2nd thought this may even be a good idea. We have some collection-array conversions in our code that implicitly copies. Not sure if there are existing utility classes that do this, I'm not aware of any at this moment.

Reply:
+1 for `UnmodifiableList `

Reply:
Considering `UnmodifiableList` is a standalone data structure that more people may need, separated that into a standalone PR https://github.com/Alluxio/alluxio/pull/15197
And therefore that PR will become a prerequisite. Pls expedite in reviewing 15197 so that I can rebase this PR.